src/assert_not_impl.rs

```diff
@@ -0,0 +1,107 @@
+/// Asserts that the type does **not** implement all of the given traits.
+///
+/// This can be used to ensure types implement auto traits such as [`Send`] and
+/// [`Sync`], as well as traits with [blanket `impl`s][blanket].
+///
+/// # Examples
+///
+/// On stable Rust, using the macro requires a unique “label” when used in a
+/// module scope:
+///
+#[cfg_attr(feature = "nightly", doc = "```ignore")]
+#[cfg_attr(not(feature = "nightly"), doc = "```")]
+/// # #[macro_use] extern crate static_assertions;
+/// # fn main() {}
+/// assert_not_impl!(str; *const u8, Sync, From<&'static str>);
+/// assert_not_impl!(vec; *const u8, Sync);
+/// ```
+///
+/// The [labeling limitation](index.html#limitations) is not necessary if
+/// compiling on nightly Rust with the `nightly` feature enabled:
+///
+#[cfg_attr(feature = "nightly", doc = "```")]
+#[cfg_attr(not(feature = "nightly"), doc = "```ignore")]
+/// #![feature(underscore_const_names)]
+/// # #[macro_use] extern crate static_assertions;
+///
+/// assert_not_impl!(&'static mut u8, Copy);
+///
+/// fn main() {
+///     assert_not_impl!(u32, Into<usize>);
+/// }
+/// ```
+///
+/// The following example fails to compile since `u32` can be converted into `u64`.
+///
+/// ```compile_fail
+/// # #[macro_use] extern crate static_assertions;
+/// # fn main() {
+/// assert_not_impl!(u32, Into<u64>);
+/// # }
+/// ```
+///
+/// Note that all traits given in a single invocation need to not be implemented. If you want to
+/// check that none of multiple trait are implemented you must invoke the macro several times.
+///
+#[cfg_attr(feature = "nightly", doc = "```ignore")]
+#[cfg_attr(not(feature = "nightly"), doc = "```")]
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // Cell<u32> is not both Sync and Send
+/// assert_not_impl!(cell; Cell<u32>, Sync, Send);
+/// ```
+///
+/// ```compile_fail
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // But it is Send, this fails to compile
+/// assert_not_impl!(cell; Cell<u32>, Send);
+/// ```
+///
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+/// [`Sync`]: https://doc.rust-lang.org/std/marker/trait.Sync.html
+/// [blanket]: https://doc.rust-lang.org/book/second-edition/ch10-02-traits.html#using-trait-bounds-to-conditionally-implement-methods
+#[macro_export(local_inner_macros)]
+macro_rules! assert_not_impl {
```

Comment:
Please rename this to `assert_not_impl_all!` and add a `assert_not_impl_any!` counterpart.

For `assert_not_impl_any!`, I strongly suggest that the implementation does:

```rust
$(impl<T: ?Sized + $t> AmbiguousIfImpl<&dyn $t> for Check<T> {})+
```